src/components/shared/ContractFunction.vue

```diff
@@ -0,0 +1,263 @@
+<script setup lang="ts">
+import useWeb3 from '@/compositions/useWeb3'
+import { decodeData, MethodType } from '@/helpers/functionUtils'
+import { reactive, computed, watch, onMounted } from 'vue'
+import { toWei, Unit, padLeft } from 'web3-utils'
+import ParamField from './ParamField.vue'
+
+interface ElementType extends MethodType {
+  error?: boolean
+}
+
+type Props = {
+  call?: string
+  modelValue?: MethodType[]
+  custom?: boolean
+  onlyParams?: boolean
+  hideData?: boolean
+  data?: string
+  testidPrefix?: string
+}
+
+type Emits = {
+  (event: 'update:modelValue', value: any): void
+  (event: 'update:error', value: boolean): void
+  (event: 'update:data', value: string): void
+}
+
+const props = defineProps<Props>()
+const emits = defineEmits<Emits>()
+
+const { getWeb3 } = useWeb3()
+const { eth } = getWeb3()
+
+function convertModel(model?: MethodType[]) {
+  model =
+    model?.map((info: MethodType) => {
+      let value = info.value ?? undefined
+      if (info.type.match(/\[\]$/) && value == null) {
+        value = []
+      }
+      return {
+        ...info,
+        value,
+        error: false,
+      }
+    }) || []
+  return model
+}
+
+const data = reactive<{ call?: string; items: ElementType[] }>({
+  call: props.call,
+  items: convertModel(props.modelValue),
+})
+
+watch(
+  () => props.modelValue,
+  model => {
+    data.items = convertModel(model)
+  }
+)
+
+function makeValue(value: string, isWei?: Unit) {
+  if (isWei) {
+    try {
+      return toWei(value, isWei)
+    } catch (err) {
+      return undefined
+    }
+  }
+  return value
+}
+
+function handleValue(param: number, value: any) {
+  const item = data.items[param]
+  item.value = value
+  emits('update:modelValue', data.items)
+}
+
+function handleError(param: number, error?: boolean) {
+  const item = data.items[param]
+  if (item.error !== error) {
+    item.error = error
+    emits(
+      'update:error',
+      data.items.some(({ error }) => error)
+    )
+  }
+}
+
+function handleUnit(param: number, isWei?: Unit) {
+  const item = data.items[param]
+  if (isWei) {
+    item.isWei = isWei
+  } else {
+    delete item.isWei
+  }
+}
+
+function handleIsKey(param: number, isKey?: boolean) {
+  const item = data.items[param]
+  if (isKey != null) {
+    item.isKey = isKey
+  } else {
+    delete item.isKey
+  }
+}
+
+const call = computed<string>(() => {
+  if (!data.call) {
+    return ''
+  }
+  return `${data.call}(${data.items
+    ?.map(({ name, type }) => `${type} ${name}`)
+    .join(', ')})`
+})
+
+function handleCall(e: Event) {
+  const { value } = e.target as HTMLTextAreaElement
+  const [_all, newCall, newArgs] =
+    /^([a-z_]*)\(([^)]*)\)$/i.exec(value || '') || []
```

Comment:
I put them into consts. This basically parses newCall(newArgs...) so that init(bytes32 a, bytes b) would end up with newCall = 'init' and newArgs = 'bytes32 a, bytes b'

Reply:
Explained